In [27]:
## Current path display

import os
current_path = os.getcwd()
print(current_path)

d:\05_Projekte\04_Research\08_ManuBrain\01_Lab\Demo-Cell-Classification\02_Preprocessing


In [28]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from datetime import datetime
from datetime import timedelta  

## Functions 

In [29]:
# return tag from data
def fc_tag (data,index):
    return data.at[index,'d'].split('=')[2] # [2] equivalent a tag column data en txt file

# update tag in main table
def fc_tag_update(data, tag):
    for i in range(len(tag)):
        i = i + 1 
        data = data.replace({'PrimaryKey':i}, fc_tag(tag,i), regex = False)
    return data

#Sync TimeStamp

def fc_sync_timestamp2(data,ServerTimeStamp,delta_time): #Requires df, column-name and delta-time
    for date in data[ServerTimeStamp]:
        source = datetime.fromisoformat(str(date).replace('Z','+00:00'))
        date_sync = source - delta_time
        data[ServerTimeStamp] = date_sync
    return data

def fc_sync_timestamp(data,ServerTimeStamp,delta_time): #Requires df, column-name and delta-time
    for i in range(len(data)): 
        source = datetime.fromisoformat(str(data.at[i,ServerTimeStamp]).replace('Z','+00:00'))
        date_sync = source - delta_time
        data.at[i,ServerTimeStamp] = date_sync
    return data

## Paths from data set

In [30]:
date = 201027
uc_num = 2

path = str(date) + '/use_case' + str(uc_num)

# Conveyor_belt
path_data_conv = '../01_Data/' + path + '/Conveyor Speed/DataLogger.csv'
path_tag_conv = '../01_Data/' + path + '/Conveyor Speed/tagconfig.txt'

# Gate
path_data_gate = '../01_Data/' + path + '/Gate/DataLogger.csv'
path_tag_gate = '../01_Data/' + path + '/Gate/tagconfig.txt'

# Quality_Control
path_data_qc = '../01_Data/' + path + '/Quality Control/DataLogger.csv'
path_tag_qc = '../01_Data/' + path + '/Quality Control/tagconfig.txt'

# Robot_Position
path_data_rp = '../01_Data/' + path + '/Robot Position/DataLogger.csv'
path_tag_rp = '../01_Data/' + path + '/Robot Position/tagconfig.txt'

# Store
path_data_st = '../01_Data/' + path + '/Store/DataLogger.csv'
path_tag_st = '../01_Data/' + path + '/Store/tagconfig.txt'

# Flag
path_data_flag = '../01_Data/' + path + '/Flag/DataLogger.csv'
path_tag_flag = '../01_Data/' + path + '/Flag/tagconfig.txt'

# Processed data

path_data_save = '../01_Data/' + path + '/Processed/'
path_sample_save = '../01_Data/' + path + '/Processed/Samples/'


## Import data from csv

In [31]:
# Conveyor_belt
data_conv =  pd.read_csv(path_data_conv, delimiter = ',')
#df_conv = pd.DataFrame(pd.read_csv(path_data_conv, delimiter = ','))

# Gate
data_gate =  pd.read_csv(path_data_gate, delimiter = ',')
#df_gate = pd.DataFrame(pd.read_csv(path_data_gate, delimiter = ','))

# Quality_Control
data_qc =  pd.read_csv(path_data_qc, delimiter = ',')
#df_qc = pd.DataFrame(pd.read_csv(path_data_qc, delimiter = ','))

# Robot_Position
data_rp =  pd.read_csv(path_data_rp, delimiter = ',')
#df_rp = pd.DataFrame(pd.read_csv(path_data_rp, delimiter = ','))

# Store
data_st =  pd.read_csv(path_data_st, delimiter = ',')
#df_st = pd.DataFrame(pd.read_csv(path_data_st, delimiter = ','))

# Flag
data_flag =  pd.read_csv(path_data_flag, delimiter = ',')
#df_flag = pd.DataFrame(pd.read_csv(path_data_st, delimiter = ','))

## Import data tags

In [32]:
# Conveyor_belt
tag_conv =  pd.read_fwf(path_tag_conv, sep=" ", header=None, names=['a','b','c','d'])
tag_conv = tag_conv.drop([0]) # delete first row (trash)

# Gate
tag_gate =  pd.read_fwf(path_tag_gate, sep=" ", header=None, names=['a','b','c','d'])
tag_gate = tag_gate.drop([0]) # delete first row (trash)

# Quality_Control
tag_qc =  pd.read_fwf(path_tag_qc, sep=" ", header=None, names=['a','b','c','d'])
tag_qc = tag_qc.drop([0]) # delete first row (trash)

# Robot_Position
tag_rp =  pd.read_fwf(path_tag_rp, sep=" ", header=None, names=['a','b','c','d'])
tag_rp = tag_rp.drop([0]) # delete first row (trash)

# Robot_Position
tag_st =  pd.read_fwf(path_tag_st, sep=" ", header=None, names=['a','b','c','d'])
tag_st = tag_st.drop([0]) # delete first row (trash)

# Flag
tag_flag =  pd.read_fwf(path_tag_flag, sep=" ", header=None, names=['a','b','c','d'])
tag_flag = tag_flag.drop([0]) # delete first row (trash)

## Update tags

In [33]:
# Conveyor_belt
data_conv = fc_tag_update(data_conv, tag_conv)

# Gate
data_gate = fc_tag_update(data_gate, tag_gate)

# Quality_Control
data_qc = fc_tag_update(data_qc, tag_qc)

# Robot_Position
data_rp = fc_tag_update(data_rp, tag_rp)

# Store
data_st = fc_tag_update(data_st, tag_st)

# Flag
data_flag = fc_tag_update(data_flag, tag_flag)

## Syncronize ServerTimeStamp

In [34]:
delta_time_PLC = timedelta(seconds=67)
delta_time_SIM = timedelta(seconds=0)

In [35]:
data_conv = fc_sync_timestamp(data_conv,'ServerTimeStamp',delta_time_PLC)
data_conv

PrimaryKey DataType  Value  StatusCode           SourceTimeStamp  \
0         27   0x0004    300  0x00000000  2020-10-27T14:31:56.627Z   
1         28   0x0004    363  0x00000000  2020-10-27T14:31:56.627Z   
2         29   0x0004    380  0x00000000  2020-10-27T14:31:56.630Z   

                    ServerTimeStamp  
0  2020-10-27 14:30:49.627000+00:00  
1  2020-10-27 14:30:49.627000+00:00  
2  2020-10-27 14:30:49.630000+00:00

In [36]:
data_gate = fc_sync_timestamp(data_gate,'ServerTimeStamp',delta_time_SIM)
data_gate

PrimaryKey DataType Value  StatusCode           SourceTimeStamp  \
0          Speed   0x000b     0  0x00000000  2020-10-27T14:30:56.854Z   
1          Speed   0x000b     0  0x00000000  2020-10-27T14:30:57.155Z   
2           Safe   0x0001  true  0x00000000  2020-10-27T14:30:56.854Z   
3           Safe   0x0001  true  0x00000000  2020-10-27T14:30:57.155Z   
4       Position   0x000b   0.3  0x00000000  2020-10-27T14:30:56.855Z   
...          ...      ...   ...         ...                       ...   
12871      Speed   0x000b     0  0x00000000  2020-10-27T14:52:24.155Z   
12872       Safe   0x0001  true  0x00000000  2020-10-27T14:52:23.855Z   
12873       Safe   0x0001  true  0x00000000  2020-10-27T14:52:24.155Z   
12874   Position   0x000b   0.2  0x00000000  2020-10-27T14:52:23.855Z   
12875   Position   0x000b   0.2  0x00000000  2020-10-27T14:52:24.155Z   

                        ServerTimeStamp  
0      2020-10-27 14:30:56.854000+00:00  
1      2020-10-27 14:30:57.155000+00:00  
2      2020-10-27 14:30:56.854000+00:00  
3      2020-10-27 14:30:57.155000+00:00  
4      2020-10-27 14:30:56.855000+00:00  
...                                 ...  
12871  2020-10-27 14:52:24.155000+00:00  
12872  2020-10-27 14:52:23.855000+00:00  
12873  2020-10-27 14:52:24.155000+00:00  
12874  2020-10-27 14:52:23.855000+00:00  
12875  2020-10-27 14:52:24.155000+00:00  

[12876 rows x 6 columns]

In [37]:
data_qc = fc_sync_timestamp(data_qc,'ServerTimeStamp',delta_time_SIM)
data_qc

PrimaryKey DataType     Value  StatusCode           SourceTimeStamp  \
0      DistanceAbs   0x000b  4.918553  0x00000000  2020-10-27T14:30:58.704Z   
1      DistanceAbs   0x000b  4.918553  0x00000000  2020-10-27T14:30:59.004Z   
2        DistanceX   0x000b  2.385984  0x00000000  2020-10-27T14:30:58.704Z   
3        DistanceX   0x000b  2.385984  0x00000000  2020-10-27T14:30:59.004Z   
4        DistanceY   0x000b  4.301075  0x00000000  2020-10-27T14:30:58.704Z   
...            ...      ...       ...         ...                       ...   
17167   ResultCode   0x0006  2.000000  0x00000000  2020-10-27T14:52:26.004Z   
17168  DistanceAbs   0x000b  0.000000  0x00000000  2020-10-27T14:52:26.304Z   
17169    DistanceX   0x000b  0.000000  0x00000000  2020-10-27T14:52:26.304Z   
17170    DistanceY   0x000b  0.000000  0x00000000  2020-10-27T14:52:26.304Z   
17171   ResultCode   0x0006  2.000000  0x00000000  2020-10-27T14:52:26.304Z   

                        ServerTimeStamp  
0      2020-10-27 14:30:58.704000+00:00  
1      2020-10-27 14:30:59.004000+00:00  
2      2020-10-27 14:30:58.704000+00:00  
3      2020-10-27 14:30:59.004000+00:00  
4      2020-10-27 14:30:58.704000+00:00  
...                                 ...  
17167  2020-10-27 14:52:26.004000+00:00  
17168  2020-10-27 14:52:26.304000+00:00  
17169  2020-10-27 14:52:26.304000+00:00  
17170  2020-10-27 14:52:26.304000+00:00  
17171  2020-10-27 14:52:26.304000+00:00  

[17172 rows x 6 columns]

In [38]:
data_rp = fc_sync_timestamp(data_rp,'ServerTimeStamp',delta_time_PLC)
data_rp

PrimaryKey DataType       Value  StatusCode           SourceTimeStamp  \
0            527   0x0007  1042334043  0x00000000  2020-10-27T14:32:07.629Z   
1            527   0x0007  1046623377  0x00000000  2020-10-27T14:32:07.879Z   
2            528   0x0007  3190235282  0x00000000  2020-10-27T14:32:07.632Z   
3            528   0x0007  3190237876  0x00000000  2020-10-27T14:32:07.882Z   
4            529   0x0007  1054414758  0x00000000  2020-10-27T14:32:07.634Z   
...          ...      ...         ...         ...                       ...   
15333        527   0x0007  3182800615  0x00000000  2020-10-27T14:53:34.872Z   
15334        528   0x0007  3190239826  0x00000000  2020-10-27T14:53:34.634Z   
15335        528   0x0007  3190240997  0x00000000  2020-10-27T14:53:34.874Z   
15336        529   0x0007  1054419631  0x00000000  2020-10-27T14:53:34.639Z   
15337        529   0x0007  1054419222  0x00000000  2020-10-27T14:53:34.874Z   

                        ServerTimeStamp  
0      2020-10-27 14:31:00.629000+00:00  
1      2020-10-27 14:31:00.879000+00:00  
2      2020-10-27 14:31:00.632000+00:00  
3      2020-10-27 14:31:00.882000+00:00  
4      2020-10-27 14:31:00.634000+00:00  
...                                 ...  
15333  2020-10-27 14:52:27.872000+00:00  
15334  2020-10-27 14:52:27.634000+00:00  
15335  2020-10-27 14:52:27.874000+00:00  
15336  2020-10-27 14:52:27.639000+00:00  
15337  2020-10-27 14:52:27.874000+00:00  

[15338 rows x 6 columns]

In [39]:
data_st = fc_sync_timestamp(data_st,'ServerTimeStamp',delta_time_SIM)
data_st

PrimaryKey DataType  Value  StatusCode           SourceTimeStamp  \
0      DiscSize   0x0003      1  0x00000000  2020-10-27T14:31:02.469Z   
1      DiscSize   0x0003      1  0x00000000  2020-10-27T14:31:02.769Z   
2      DiscSize   0x0003      1  0x00000000  2020-10-27T14:31:03.069Z   
3      DiscSize   0x0003      1  0x00000000  2020-10-27T14:31:03.369Z   
4      DiscSize   0x0003      1  0x00000000  2020-10-27T14:31:03.669Z   
...         ...      ...    ...         ...                       ...   
4289   DiscSize   0x0003      2  0x00000000  2020-10-27T14:52:29.169Z   
4290   DiscSize   0x0003      2  0x00000000  2020-10-27T14:52:29.469Z   
4291   DiscSize   0x0003      2  0x00000000  2020-10-27T14:52:29.769Z   
4292   DiscSize   0x0003      2  0x00000000  2020-10-27T14:52:30.069Z   
4293   DiscSize   0x0003      2  0x00000000  2020-10-27T14:52:30.369Z   

                       ServerTimeStamp  
0     2020-10-27 14:31:02.469000+00:00  
1     2020-10-27 14:31:02.769000+00:00  
2     2020-10-27 14:31:03.069000+00:00  
3     2020-10-27 14:31:03.369000+00:00  
4     2020-10-27 14:31:03.669000+00:00  
...                                ...  
4289  2020-10-27 14:52:29.169000+00:00  
4290  2020-10-27 14:52:29.469000+00:00  
4291  2020-10-27 14:52:29.769000+00:00  
4292  2020-10-27 14:52:30.069000+00:00  
4293  2020-10-27 14:52:30.369000+00:00  

[4294 rows x 6 columns]

In [40]:
data_flag = fc_sync_timestamp(data_flag,'ServerTimeStamp',delta_time_PLC)
data_flag

PrimaryKey DataType  Value  StatusCode           SourceTimeStamp  \
0          561   0x0001  false  0x00000000  2020-10-27T14:32:01.126Z   
1          560   0x0004      1  0x00000000  2020-10-27T14:32:01.129Z   
2          560   0x0004      2  0x00000000  2020-10-27T14:32:02.379Z   
3          561   0x0001   true  0x00000000  2020-10-27T14:32:04.378Z   
4          560   0x0004      3  0x00000000  2020-10-27T14:32:09.129Z   
..         ...      ...    ...         ...                       ...   
371        561   0x0001   true  0x00000000  2020-10-27T14:52:12.626Z   
372        560   0x0004      3  0x00000000  2020-10-27T14:52:17.380Z   
373        561   0x0001  false  0x00000000  2020-10-27T14:52:18.127Z   
374        560   0x0004      0  0x00000000  2020-10-27T14:52:40.127Z   
375        561   0x0001   true  0x00000000  2020-10-27T14:52:40.878Z   

                      ServerTimeStamp  
0    2020-10-27 14:30:54.126000+00:00  
1    2020-10-27 14:30:54.129000+00:00  
2    2020-10-27 14:30:55.379000+00:00  
3    2020-10-27 14:30:57.378000+00:00  
4    2020-10-27 14:31:02.129000+00:00  
..                                ...  
371  2020-10-27 14:51:05.626000+00:00  
372  2020-10-27 14:51:10.380000+00:00  
373  2020-10-27 14:51:11.127000+00:00  
374  2020-10-27 14:51:33.127000+00:00  
375  2020-10-27 14:51:33.878000+00:00  

[376 rows x 6 columns]

## Merge Data Sets

In [41]:
data_set = [data_conv, data_gate, data_qc, data_rp, data_st, data_flag]
data_merge = pd.concat(data_set, keys=['conv', 'gate', 'qc', 'rp', 'store', 'flag'])
# Sort the data_set by SourceTimeStamp, reset index and delete aux column
data_merge = data_merge.sort_values(by='ServerTimeStamp').reset_index().drop('level_1',1)
#del data_merge['level_1'] #delete aux column

In [42]:
data_merge

level_0 PrimaryKey DataType  Value  StatusCode  \
0        conv         27   0x0004    300  0x00000000   
1        conv         28   0x0004    363  0x00000000   
2        conv         29   0x0004    380  0x00000000   
3        flag        561   0x0001  false  0x00000000   
4        flag        560   0x0004      1  0x00000000   
...       ...        ...      ...    ...         ...   
50054   store   DiscSize   0x0003      2  0x00000000   
50055   store   DiscSize   0x0003      2  0x00000000   
50056   store   DiscSize   0x0003      2  0x00000000   
50057   store   DiscSize   0x0003      2  0x00000000   
50058   store   DiscSize   0x0003      2  0x00000000   

                SourceTimeStamp                   ServerTimeStamp  
0      2020-10-27T14:31:56.627Z  2020-10-27 14:30:49.627000+00:00  
1      2020-10-27T14:31:56.627Z  2020-10-27 14:30:49.627000+00:00  
2      2020-10-27T14:31:56.630Z  2020-10-27 14:30:49.630000+00:00  
3      2020-10-27T14:32:01.126Z  2020-10-27 14:30:54.126000+00:00  
4      2020-10-27T14:32:01.129Z  2020-10-27 14:30:54.129000+00:00  
...                         ...                               ...  
50054  2020-10-27T14:52:29.169Z  2020-10-27 14:52:29.169000+00:00  
50055  2020-10-27T14:52:29.469Z  2020-10-27 14:52:29.469000+00:00  
50056  2020-10-27T14:52:29.769Z  2020-10-27 14:52:29.769000+00:00  
50057  2020-10-27T14:52:30.069Z  2020-10-27 14:52:30.069000+00:00  
50058  2020-10-27T14:52:30.369Z  2020-10-27 14:52:30.369000+00:00  

[50059 rows x 7 columns]

## Export Data to csv

In [43]:
data_merge.to_csv(path_data_save + 'data_merge.csv')
print('The data is saved at ' + path_data_save + 'data_merge.csv')

The data is saved at ../01_Data/201027/use_case2/Processed/data_merge.csv


In [44]:
len(data_merge)

50059

## Determine the starting point of data set

#Based on change of DistanceAbs in Data- Set
var_actual = 0
run_list = [0]
for i in range(len(data_merge)):
    var_previous = var_actual
    if data_merge.at[i,'PrimaryKey'] == 'DistanceAbs':  # if 'PrimaryKey' value = 'DistanceAbs':
        var_actual = data_merge.at[i,'Value'] # value of current DistanceAbs
    if var_previous != var_actual:
        run_list.append(i) #i append to list # This list will tell at how many i contains one sample    

# Based on grabbing the white disk
var_actual = 99
disk_actual = 99 # Initial value
run_list = [0]
disk = False
storage_down = False
rob_x = 0
rob_y = 0
rob_z = 0
rob_pos = 0
pos = [0]*(len(data_merge)-2)
for i in range(len(data_merge)-2): # len(data_merge)-2 is considered to avoid border errors
    #var_previous = var_actual
    #rob_pos = fc_rob_pos(i, data_merge) # review robot position
    pos[i] = 0
    if data_merge.at[i,'PrimaryKey'] == 'DiscSize':  # if 'PrimaryKey' value = 'DiskSize':
        disk_previous = disk_actual
        disk_actual = data_merge.at[i,'Value'] # value of current DiskSize
        #print(disk_previous, disk_actual, i)
        if disk_previous != 2 and disk_actual == 2: # Varialbe 0: unkown; 1: Black disk; 2: white disk
            disk = True
            #print('New white disk here ' + str(i))
    if data_merge.at[i,'PrimaryKey'] == '527':  # if 'PrimaryKey' value = 'x coordinate':
        rob_x = data_merge.at[i,'Value'] # value of current 'x coordinate'
        #print('x = ' + str(rob_x) + '--' + str(i))
    if data_merge.at[i,'PrimaryKey'] == '528':  # if 'PrimaryKey' value = 'y coordinate':
        rob_y = data_merge.at[i,'Value'] # value of current 'x coordinate'
        #print('y = ' + str(rob_y) + '--' + str(i))
    if data_merge.at[i,'PrimaryKey'] == '529':  # if 'PrimaryKey' value = 'z coordinate':
        rob_z = data_merge.at[i,'Value'] # value of current 'x coordinate'
        #print('z = ' + str(rob_z) + '--' + str(i))
    if rob_x > 3100000000 and rob_y > 3100000000 and rob_z < 1051200000: #Robot in storage. Picking up the disk
        rob_pos = 1
        pos[i] = 1
        #print('pos 1 - ' + str(i))
    if rob_x < 1052000000 and rob_y > 3100000000 and rob_z < 1052000000 and rob_z > 1051200000 : #Robot in storage. Placing down the disk
        rob_pos = 2
        pos[i] = 2
        #print('pos 2 - ' + str(i))
    #print(rob_pos)
    if rob_pos == 1: # Robot picking up a disk
        storage_down = True
        #print(storage_down,i)  
    if disk == True and storage_down == True:
        #var_actual = rob_pos
        disk = False
        storage_down = False
        run_list.append(i) #i append to list # This list will tell at how many i contains one sample
        #print(i)
    #if var_previous != var_actual:
        #run_list.append(i) #i append to list # This list will tell at how many i contains one sample   
 

In [45]:
# Based on grabbing the white disk
run_list = [0]
rob_pos_actual = 99
rob_pos_old = 0
gripper_actual = 99
gripper_old = 0
gripper = False
rob_pos = False
for i in range(len(data_merge)-2): # len(data_merge)-2 is considered to avoid border errors
    #var_previous = var_actual
    #rob_pos = fc_rob_pos(i, data_merge) # review robot position
    if data_merge.at[i,'PrimaryKey'] == '561':  # if 'PrimaryKey' value = 'gripper open (0), close (1)':
        gripper_old = gripper_actual
        gripper_actual = data_merge.at[i,'Value'] # value of current 'x coordinate'
        if gripper_old != gripper_actual:
            gripper = True
    if data_merge.at[i,'PrimaryKey'] == '560':  # if 'PrimaryKey' value = 'position of robot based in certain conditions':
        # 0: Black disk in store
        # 1: Black disk in gate
        # 2: White disk in store
        # 3: White disk in gate
        rob_pos_old = rob_pos_actual
        rob_pos_actual = int(data_merge.at[i,'Value']) # value of current 'x coordinate'
        if rob_pos_old != rob_pos_actual and rob_pos_actual == 2:
            rob_pos = True
    if rob_pos == True and gripper == True:
        rob_pos = False
        gripper = False
        run_list.append(i) #i append to list # This list will tell at how many i contains one sample
        #print(i)
 

In [46]:
print(run_list) # Starting points. 
print(len(run_list))

[0, 5, 886, 1909, 2933, 3955, 4989, 6009, 7035, 8055, 9082, 10101, 11127, 12138, 13161, 14184, 15207, 16230, 17253, 18276, 19311, 20331, 21356, 22377, 23403, 24422, 25448, 26468, 27494, 28514, 29540, 30561, 31597, 32617, 33643, 34664, 35689, 36719, 37735, 38765, 39781, 40811, 41833, 42856, 43887, 44913, 45926, 46948]
48


In [47]:
# data_merge.iloc[list(range(0,3))]

In [48]:
#df_robpos = pd.DataFrame(pos,columns=['RobPos']) # Position of robot. 0: not important; 1: Robot in storage. Picking up the disk; 2:Robot in storage. Placing down the disk
#data_merge = pd.concat([data_merge, df_robpos], axis=1, sort=False)
data_merge

level_0 PrimaryKey DataType  Value  StatusCode  \
0        conv         27   0x0004    300  0x00000000   
1        conv         28   0x0004    363  0x00000000   
2        conv         29   0x0004    380  0x00000000   
3        flag        561   0x0001  false  0x00000000   
4        flag        560   0x0004      1  0x00000000   
...       ...        ...      ...    ...         ...   
50054   store   DiscSize   0x0003      2  0x00000000   
50055   store   DiscSize   0x0003      2  0x00000000   
50056   store   DiscSize   0x0003      2  0x00000000   
50057   store   DiscSize   0x0003      2  0x00000000   
50058   store   DiscSize   0x0003      2  0x00000000   

                SourceTimeStamp                   ServerTimeStamp  
0      2020-10-27T14:31:56.627Z  2020-10-27 14:30:49.627000+00:00  
1      2020-10-27T14:31:56.627Z  2020-10-27 14:30:49.627000+00:00  
2      2020-10-27T14:31:56.630Z  2020-10-27 14:30:49.630000+00:00  
3      2020-10-27T14:32:01.126Z  2020-10-27 14:30:54.126000+00:00  
4      2020-10-27T14:32:01.129Z  2020-10-27 14:30:54.129000+00:00  
...                         ...                               ...  
50054  2020-10-27T14:52:29.169Z  2020-10-27 14:52:29.169000+00:00  
50055  2020-10-27T14:52:29.469Z  2020-10-27 14:52:29.469000+00:00  
50056  2020-10-27T14:52:29.769Z  2020-10-27 14:52:29.769000+00:00  
50057  2020-10-27T14:52:30.069Z  2020-10-27 14:52:30.069000+00:00  
50058  2020-10-27T14:52:30.369Z  2020-10-27 14:52:30.369000+00:00  

[50059 rows x 7 columns]

The list contains the point i when the value is different. To consider the point where the value remains, consider i-1

In [49]:
d = {} # Dictionary that holds the sub-data-sets given by the starting point
for i in range(len(run_list)-1): # Ignores the last value (set of values)
    list_range = list(range(run_list[i],run_list[i+1]))
    d['df{0}'.format(i)] = data_merge.iloc[list_range]
    
print('Dictionary lenght is:' + str(len(d)))

Dictionary lenght is:47


## Assure conv_speed per sample 

In [50]:
conv_1 = 0
conv_2 = 0
conv_3 = 0
conv1 = False
conv2 = False
conv3 = False
for j in range(len(d)): # as many samples saved in dictionary
    df = d['df'+ str(j)]
    for i in range(len(df)): # as many row in the dataFrame
        #print(df.iat[i,1])
        if df.iat[i,1] == '27': # conv1
            conv_1 = df.iat[i,3]
            conv1 = True # the value shows at least once in the data set
        if df.iat[i,1] == '28': # conv2
            conv_2 = df.iat[i,3]
            conv2 = True
        if df.iat[i,1] == '29': # conv3
            conv_3 = df.iat[i,3]
            conv3 = True
    if conv1 == False: # The value did not occur in this sample. Copy value from previous samples into this one
        dic = {'PrimaryKey' : '27', 'Value': str(conv_1)}
        df.loc[len(df)+1,'PrimaryKey'] = '27'
        df.loc[len(df),'Value'] = str(conv_1)
        df.loc[len(df),'DataType'] = 'NA'
        df.loc[len(df),'StatusCode'] = 'NA'
        df.loc[len(df),'SourceTimeStamp'] = 'NA'
        df.loc[len(df),'ServerTimeStamp'] = 'NA'
    if conv2 == False:
        dic = {'PrimaryKey' : '27', 'Value': str(conv_2)}
        df.loc[len(df)+1,'PrimaryKey'] = '28'
        df.loc[len(df),'Value'] = str(conv_2)
        df.loc[len(df),'DataType'] = 'NA'
        df.loc[len(df),'StatusCode'] = 'NA'
        df.loc[len(df),'SourceTimeStamp'] = 'NA'
        df.loc[len(df),'ServerTimeStamp'] = 'NA'
    if conv3 == False:
        dic = {'PrimaryKey' : '27', 'Value': str(conv_3)}
        df.loc[len(df)+1,'PrimaryKey'] = '29'
        df.loc[len(df),'Value'] = str(conv_3)
        df.loc[len(df),'DataType'] = 'NA'
        df.loc[len(df),'StatusCode'] = 'NA'
        df.loc[len(df),'SourceTimeStamp'] = 'NA'
        df.loc[len(df),'ServerTimeStamp'] = 'NA'
   
    conv1 = False # reset conv for the next round
    conv2 = False
    conv3 = False
print(conv_1)
print(conv_2)
print(conv_3)

300
363
380


## Export individual data-sample to .csv

In [51]:
for i in range(len(d)):
    if i > 0: # always ignore the first sample due to data-acquisition misalligment
        d['df' + str(i)].to_csv(path_sample_save + 'sample_' + str(i) + '.csv')
print('The data is saved at ' + path_sample_save + 'sample_x.csv')

The data is saved at ../01_Data/201027/use_case2/Processed/Samples/sample_x.csv


## Testing area

In [52]:
#d['df67']

In [53]:
# list_range = list(range(run_list[0],run_list[1]))
# print(list_range)

In [54]:
# data_merge.iloc[list_range]

In [55]:
# data_rp.loc[:, 'PrimaryKey'] = '529_z'
# data_rp

In [56]:
# %matplotlib notebook 
# data_gate.head(400).plot(x='SourceTimeStamp', y= 'Value')
# plt.xticks(rotation=45)

In [57]:
# %matplotlib notebook 
# data_rp.head(200).plot( y= 'Value')
# plt.xticks(rotation=45)